### PR TITLE
feat(core): introduce TOPK data structure (Stage 1)

### PR DESCRIPTION
**Pull request by glevkovich**
- Introduce the core TOPK class to provide the backing data structure for the Redis Top-K command family.
- Add topk.cc to dfly_core and implement a thorough suite of unit tests (topk_test) covering all edge cases and math invariants.
- Separated this from the internal TopKeys tracker to provide PMR support and RDB serialization required for public Redis APIs.

### CODE REVIEW

File: src/core/topk.h [hunk 1]
```diff
@@ -0,0 +1,269 @@
...
+#pragma once
+
+#include <cstdint>
+#include <memory>
+#include <optional>
+#include <queue>
+#include <string>
+#include <string_view>
+#include <vector>
+
```
[the remark this answers is not in the document]
Comment (glevkovich) [reply src/core/topk.h:15]: fixed

File: src/core/topk_test.cc [hunk 2]
```diff
@@ -0,0 +1,729 @@
...
+// Counter Saturation (Overflow Prevention)
+// ---------------------------------------------------------------------------
+
+// IncrBy must saturate at UINT32_MAX rather than wrapping around to 0.
+// A wrap-around would trick the heap into evicting a top item — a correctness
+// and security issue (malicious TOPK.INCRBY with a huge increment).
+TEST_F(TOPKTest, CounterSaturationPreventsOverflow) {
+  const uint32_t max_val = numeric_limits<uint32_t>::max();
+  topk_.IncrBy("max_item", max_val);
+  EXPECT_EQ(topk_.Count({"max_item"})[0], max_val);
```
[the remark this answers is not in the document]
Comment (glevkovich) [reply src/core/topk_test.cc]: fixed

File: src/core/topk.cc [hunk 3]
```diff
@@ -0,0 +1,440 @@
...
+
+  return data;
+}
+
+void TOPK::Deserialize(const SerializedData& data) {
+  DCHECK_EQ(data.counters.size(), static_cast<size_t>(width_) * depth_);
+  DCHECK_LE(data.heap_items.size(), k_);
+  DCHECK_EQ(data.k, k_);
+  DCHECK_EQ(data.width, width_);
+  DCHECK_EQ(data.depth, depth_);
```
[the remark this answers is not in the document]
Comment (glevkovich) [reply src/core/topk.cc:355]: fixed

File: src/core/topk.cc [hunk 4]
```diff
@@ -0,0 +1,440 @@
+// Copyright 2026, DragonflyDB authors.  All rights reserved.
+// See LICENSE for licensing terms.
+//
+
+#include "core/topk.h"
+
+#include <xxhash.h>
+
+#include <algorithm>
+#include <cmath>
```
[the remark this answers is not in the document]
Comment (glevkovich) [reply src/core/topk.cc:10]: fixed

File: src/core/topk.h [hunk 5]
```diff
@@ -0,0 +1,269 @@
...
+  // Updates the min-heap with the new count for the given item.
+  // Returns the evicted item's key if the heap is at capacity and a new item displaces an existing
+  // one. Otherwise, returns std::nullopt.
+  std::optional<std::string> UpdateHeap(std::string_view item, uint32_t new_count);
+
+  // Check if an item is in the Top-K heap
+  [[nodiscard]] bool IsInHeap(std::string_view item) const {
+    return item_to_hash_.contains(item);
+  }
```
[the remark this answers is not in the document]
Comment (glevkovich) [reply src/core/topk.h:201]: this path is already optmized by absl. dropping.

File: src/core/topk.h [hunk 6]
```diff
@@ -0,0 +1,269 @@
...
+
+  // Extracts the current structural state of the sketch for RDB persistence.
+  [[nodiscard]] SerializedData Serialize() const;
+
+  // Reconstructs the internal state of the sketch from a previously serialized dataset.
+  void Deserialize(const SerializedData& data);
+
+ private:
+  struct HeapItem {
+    std::string key;
```
[the remark this answers is not in the document]
Comment (glevkovich) [reply src/core/topk.h:174]: will be done on next stage

File: src/core/top_keys.h [hunk 7]
```diff
@@ -9,10 +9,20 @@
 #include <string>
 #include <string_view>
 #include <vector>
+
 #include "base/random.h"
 
 namespace dfly {
 
+// INTERNAL USE ONLY: This class is an optimized, O(1) probabilistic hot-key tracker designed
+// specifically to run on the database's hot path (e.g., tracking hot keys for DEBUG HTKEYS).
```
Comment (romange) [inline src/core/top_keys.h]: HTKEYS -> `TOPK`

File: src/core/topk.cc [hunk 8]
```diff
@@ -0,0 +1,441 @@
...
+#include "absl/random/distributions.h"
+#include "base/logging.h"
+#include "base/random.h"
+
+namespace dfly {
+
+namespace {
+
+const std::array<double, TOPK::kDecayLookupSize>& GetDefaultDecayTable() {
+  static const auto table = [] {
```
Comment (romange) [inline src/core/topk.cc:23]: can you please explain?


Image

File: src/core/topk.cc [hunk 9]
```diff
@@ -0,0 +1,441 @@
...
+            [](const TopKItem& a, const TopKItem& b) { return a.count > b.count; });
+
+  return result;
+}
+
+std::optional<std::string> TOPK::UpdateHeap(std::string_view item, uint32_t new_count) {
+  // Fast path: O(1) hash lookup for membership, followed by a cache-friendly
+  // O(K) linear scan and O(log K) heap update.
+  auto it = item_to_hash_.find(item);
+  if (it != item_to_hash_.end()) {
```
Comment (romange) [inline src/core/topk.cc]: it should not happen?

File: src/core/topk.cc [hunk 10]
```diff
@@ -0,0 +1,441 @@
...
+  uint32_t min_count = GetMinCount(item);
+
+  return UpdateHeap(item, min_count);
+}
+
+std::optional<std::string> TOPK::Add(std::string_view item) {
+  return IncrementInternal(item, 1);
+}
+
+std::vector<std::optional<std::string>> TOPK::AddMultiple(
```
Comment (romange) [inline src/core/topk.cc]: why this should be part of the  API?

File: src/core/topk.h [hunk 11]
```diff
@@ -0,0 +1,276 @@
...
+  [[nodiscard]] bool IsInHeap(std::string_view item) const {
+    return item_to_hash_.contains(item);
+  }
+
+  // Hashes the item for a specific row and calculates its flattened 1D index
+  // within the counters_ array. Maps the 2D Count-Min Sketch grid (depth x width)
+  // into a single contiguous block of memory for better CPU cache locality.
+  size_t GetCounterIndex(std::string_view item, uint32_t row) const;
+
+  // Shared increment logic
```
Comment (romange) [inline src/core/topk.h:208]: what does it return?

File: src/core/topk.h [hunk 12]
```diff
@@ -0,0 +1,276 @@
...
+  //          contains either the evicted item resulting from that specific
+  //          insertion, or std::nullopt.
+  std::vector<std::optional<std::string>> AddMultiple(const std::vector<std::string_view>& items);
+
+  // Increments an item's estimated frequency by a specific, arbitrary amount.
+  //
+  // Precondition: 'increment' must be strictly greater than 0.
+  // Returns: The string of the evicted item if this operation caused a resident
+  //          item to be displaced from the Top-K min-heap, or std::nullopt.
+  std::optional<std::string> IncrBy(std::string_view item, uint32_t increment);
```
Comment (romange) [inline src/core/topk.h:105]: do we need returning the evicted item as part of the TOPK API?

File: src/core/topk.h [hunk 13]
```diff
@@ -0,0 +1,276 @@
...
+  std::vector<std::optional<std::string>> IncrByMultiple(
+      const std::vector<std::pair<std::string_view, uint32_t>>& items);
+
+  // Queries whether a batch of items currently resides in the Top-K min-heap.
+  //
+  // Input: A vector of strings (items) to check against the Top-K list.
+  // Output: A vector of integers mapping 1:1 to the input vector.
+  //         The i-th element of the output will be 1 if items[i] is
+  //         currently a Top-K high-frequency item, and 0 otherwise.
+  [[nodiscard]] std::vector<int> Query(const std::vector<std::string_view>& items) const;
```
Comment (romange) [inline src/core/topk.h]: I think these APIs do not contribute much - I did not see any batch-related optimization that justifies this, so why?

File: src/core/top_keys.h (hunk 7 above)
Comment (glevkovich) [reply src/core/top_keys.h]: fixed

File: src/core/topk.cc (hunk 8 above)
Comment (glevkovich) [reply src/core/topk.cc:23]: I wrote explanations in header. There are two key  choices here:

- Speed vs. Memory: The lookup table pre-calculates pow() results to keep the hot path O(1). I increased its size to prioritize execution speed over a small, 32KB memory trade-off.
- Shared Defaults: I assume that vast majority of users will use the default decay. If no one users it, we wasted 32KB of memory. We share this single table across all TOPK instances using the default decay. This means 1,000 keys only occupy the memory of one table. Custom (non-default) decays allocate their own instance-specific table (which we can optimize for deduplication later).
 
Also, why the lambda and thread safety?  This is thread-safe by default due to C++11 "magic statics" . the compiler ensures the table is built exactly once, even if multiple threads call this simultaneously, without needing manual locks.

File: src/core/topk.cc (hunk 9 above)
Comment (glevkovich) [reply src/core/topk.cc]: Could you clarify your point regarding the membership check?I first used item_to_hash_ as a fast filter, assuming the O(1) lookup would be more efficient than a scan. However, I now realize this creates a 'double lookup' (hashing the string for the map, then performing a linear scan anyway to find the heap index).

Rethinking, perhaps I should remove the  outer find find() entirely.. given that K is typically small , maybe a single linear scan to benefit from better cache locality and avoid the CPU overhead of hashing the string twice.

Comment (glevkovich) [reply src/core/topk.cc]: This also simplifies code, reduce memory overhead since item_to_hash_ is removed , and simplify memory tracking for PMRs.

Comment (glevkovich) [reply src/core/topk.cc]: So aliminating this double lookup will actually improve performance. The decision to use a hash map for O(1) was not optimal , since K is typically small, the overhead of hashing and potential cache misses outweighs the benefits of the map.I will remove the item_to_hash_ and relay on a linear scan and cache locallity.

File: src/core/topk.cc (hunk 10 above)
Comment (glevkovich) [reply src/core/topk.cc]: Im going to remove  AddMultiple/IncrByMultiple - they are leftover convenience wrappers that I did not remove. , The family layer (topk_family.cc) can simply loop over and call Add/IncrBy and stream replies directly with less intermidiate vectors allocated.

File: src/core/topk.h (hunk 11 above)
Comment (glevkovich) [reply src/core/topk.h:208]: It actually calls  UpdateHeap and returns it return value. I will update the comment above.

```
// Increments the item's count in the sketch and updates the min-heap.
  // Returns: The key of the item evicted from the Top-K list, or std::nullopt 
  //          if no eviction occurred.
  std::optional<std::string> IncrementInternal(std::string_view item, uint32_t increment);
```

File: src/core/topk.h (hunk 12 above)
Comment (glevkovich) [reply src/core/topk.h:105]: Yes, it is required. The Redis API  for both TOPK.ADD and TOPK.INCRBY dictates that the command must return "Array of Simple string reply - if an element was dropped from the TopK list, Nil reply otherwise.."

File: src/core/topk.h (hunk 13 above)
Comment (glevkovich) [reply src/core/topk.h]: i;m going to change both Query and Count to have a single input element.

File: src/core/topk.cc [hunk 14]
```diff
@@ -0,0 +1,368 @@
+// Copyright 2026, DragonflyDB authors.  All rights reserved.
+// See LICENSE for licensing terms.
+//
+
+#include "core/topk.h"
+
+#include <xxhash.h>
+
+#include <algorithm>
+#include <cmath>
```
[the remark this answers is not in the document]
Comment (glevkovich) [reply src/core/topk.cc:10]: fixed

File: src/core/topk.h [hunk 15]
```diff
@@ -0,0 +1,249 @@
...
+  // Returns: The string of the evicted item if this insertion caused a resident
+  //          item to be displaced from the Top-K min-heap, or std::nullopt
+  //          if no eviction occurred.
+  std::optional<std::string> Add(std::string_view item);
+
+  // Increments an item's estimated frequency by a specific, arbitrary amount.
+  //
+  // Precondition: 'increment' must be strictly greater than 0.
+  // Returns: The string of the evicted item if this operation caused a resident
+  //          item to be displaced from the Top-K min-heap, or std::nullopt.
```
[the remark this answers is not in the document]
Comment (glevkovich) [reply src/core/topk.h]: fixed

File: src/core/topk.cc (hunk 8 above)
Comment (romange) [reply src/core/topk.cc:23]: yes, I asked about the lambda hence the inception meme.
Can you explain the "magic statics" hack? why it is thread-safe? I was not aware of it.

File: src/core/topk.cc [hunk 16]
```diff
@@ -0,0 +1,369 @@
...
+    } else {
+      counters_[idx] = static_cast<uint32_t>(
+          std::min(static_cast<uint64_t>(counters_[idx]) + increment,
+                   static_cast<uint64_t>(std::numeric_limits<uint32_t>::max())));
+    }
+  }
+
+  // Count-Min Sketch property: The minimum counter across all rows is the
+  // most accurate, as it has suffered the fewest hash collisions.
+  uint32_t min_count = GetMinCount(item);
```
Comment (romange) [inline src/core/topk.cc]: you do not need `GetMinCount` here, you can calculate it in the loop

File: src/core/topk.cc [hunk 17]
```diff
@@ -0,0 +1,369 @@
...
+  // Sort by count (descending) for output
+  std::sort(result.begin(), result.end(),
+            [](const TopKItem& a, const TopKItem& b) { return a.count > b.count; });
+
+  return result;
+}
+
+std::optional<std::string> TOPK::UpdateHeap(std::string_view item, uint32_t new_count) {
+  // Fast path: O(K) cache-friendly linear scan over contiguous memory.
+  // For small K, this is faster than paying the overhead of a hash map lookup.
```
Comment (romange) [inline src/core/topk.cc]: I am not so sure and it's not contiguous memory.`min_heap_[i].key == item` is not contiguous. 

I any case I am fine to start like this and then we can follow up with benchmarking.

File: src/core/topk.cc (hunk 16 above)
Comment (glevkovich) [reply src/core/topk.cc]: Thanks, I'll move it into the loop.

File: src/core/topk.cc [hunk 18]
```diff
@@ -0,0 +1,369 @@
...
+  }
+}
+
+void TOPK::HeapifyDown(size_t index) {
+  DCHECK_LT(index, min_heap_.size());
+  // Restores the min-heap property by shifting the element at 'index' downward.
+  // Triggered in two cases:
+  // 1. Root replacement/removal: The minimum item is evicted/replaced and the new root must sink.
+  // 2. Count increase: An existing item's count grows (becomes heavier), sinking lower.
+  size_t size = min_heap_.size();
```
Comment (romange) [inline src/core/topk.cc:166]: would https://en.cppreference.com/w/cpp/algorithm/make_heap.html 

work here?

File: src/core/topk.cc (hunk 17 above)
Comment (glevkovich) [reply src/core/topk.cc]: The std::string data pointers break contiguity for larger strings. (not for SSO) .I'll update the comment to be more accurate, and we should follow  up with benchmarking on later stages..

File: src/core/topk.cc (hunk 18 above)
Comment (glevkovich) [reply src/core/topk.cc:166]: btw.i'm using make_heap on Deserialize().

And it will work here functionally-wise, but will be a performance disaster probably  - it throws the existing heap and build it from scratch. STL provided pop_heap() and push_heap() to add item at the end or remove from the root. but no API for updating item in the middle of the heap. We are updating arbitrary indices in the heap.

**Review by romange — DISMISSED** [review]
lgtm

File: src/core/topk.cc (hunk 8 above)
Comment (glevkovich) [reply src/core/topk.cc:23]: Adding here part of the message I wrote in the C++ channel:

- Before C++11, doing this concurrently was undefined behavior.
- Since C++11, block-scope static variables (often called "magic statics") are guaranteed to be thread-safe by the compiler.
- According to cppreference: "If multiple threads attempt to initialize the same static local variable concurrently, the initialization occurs exactly once." (See the 'Static local variables' section here: https://en.cppreference.com/w/cpp/language/storage_duration).
- Because of this compiler-level guarantee, my claim is that initializing this shared lookup table using a static lambda is perfectly thread-safe and avoids the overhead of manual locking.
